Add advanced analytics tests to gpdb master pipeline

kaknikhil opened the pull request:
Add madlib tests on the master piepline, and also make those jobs as a requirements for release candidates.

jvigil-pivotal commented on concourse/pipelines/templates/gpdb-tpl.yml:
```diff
@@ -267,6 +283,21 @@ groups:
+  jobs:
+  - gate_advanced_analytics_start
+{% for test_name in aa_Test_suites %}
+{% set orca_state = 'off' if "planner" in test_name else 'on' %}
```
It seems like this line might be unnecessary because `orca_state` isn't used here -- only in the job definition.

jvigil-pivotal commented on concourse/pipelines/templates/gpdb-tpl.yml:
```diff
@@ -2124,6 +2173,110 @@ jobs:
+      trigger: true
+      passed:
+{% for test_name in aa_Test_suites %}
+{% set orca_state = 'off' if "planner" in test_name else 'on' %}
```
Same here

jvigil-pivotal commented on concourse/pipelines/templates/gpdb-tpl.yml:
```diff
@@ -2124,6 +2173,110 @@ jobs:
+    - get: bin_gpdb_centos6
+      passed:
+{% for test_name in aa_Test_suites %}
+{% set orca_state = 'off' if "planner" in test_name else 'on' %}
```
And here

jvigil-pivotal commented on concourse/pipelines/templates/gpdb-tpl.yml:
```diff
@@ -2124,6 +2173,110 @@ jobs:
+    - get: bin_gpdb_centos7
+      passed:
+{% for test_name in aa_Test_suites %}
+{% set orca_state = 'off' if "planner" in test_name else 'on' %}
```
And here

l-wang commented on concourse/pipelines/gpdb_master-generated.yml:
```diff
@@ -2145,6 +2303,10 @@ jobs:
+      - MADlib_Test_planner_centos6
+      - MADlib_Test_planner_centos7
+      - MADlib_Test_orca_centos6
+      - MADlib_Test_orca_centos7
```
these should be added to the passed jobs of `bin_gpdb_centos6` and `bin_gpdb_centos7` right below this task accordingly.

kaknikhil commented:
@l-wang @jvigil-pivotal thank you for the review. I have pushed a new commit for addressing the review comments. Can you take a look and approve the PR if everything looks good ?

kmacoskey commented on concourse/pipelines/gpdb_master-generated.yml:
```diff
@@ -202,6 +210,16 @@ groups:
   - compile_gpdb_centos6
 
+
+- name: Advanced Analytics
```
A super small thing, but could you please drop the space in the group name? The other groups that are two words don't have any spaces in the name.

kmacoskey commented on concourse/pipelines/gpdb_master-generated.yml:
```diff
@@ -2167,6 +2201,130 @@ jobs:
+
+
+
+- name: gate_advanced_analytics_end
```
Unless there is a particular use for this job that is planned (currently it just does `gets` and doesn't interact with anything nor `put` anything), the job should just be excluded. When we fan-in to determine the release candidates in the `Release_Candidate` we wouldn't be able to use the gate anyways.

kaknikhil replied:
Sure makes sense, I will remove the end gate.

kmacoskey commented on concourse/pipelines/gpdb_master-generated.yml:
```diff
@@ -2250,6 +2412,10 @@ jobs:
       - gpexpand_1
       - gpexpand_2
+      - MADlib_Test_planner_centos6
+      - MADlib_Test_planner_centos7
```
This is the `get` section for `bin_gpdb_centos6` so only `centos6` related artifacts should be included.

kmacoskey commented on concourse/pipelines/gpdb_master-generated.yml:
```diff
@@ -2263,6 +2429,10 @@ jobs:
       - icw_gporca_centos7
       - icw_planner_centos7
       - resource_group_centos7
+      - MADlib_Test_planner_centos6
```
This is the get section for `bin_gpdb_centos7` so only `centos7` related artifacts should be included.

kmacoskey commented on concourse/pipelines/templates/gpdb-tpl.yml:
```diff
@@ -114,6 +117,19 @@ groups:
+  ## --------------------------------------------------------------------
+  - gate_advanced_analytics_start
+{% for test_name in aa_Test_suites %}
+{% set orca_state = 'off' if "planner" in test_name else 'on' %}
```
The `orca_state` doesn't actually effect anything here so this line should be removed.

kmacoskey commented on concourse/pipelines/templates/gpdb-tpl.yml:
```diff
@@ -2198,6 +2246,107 @@ jobs:
+{% endif %}
+
+{% for test_name in aa_Test_suites %}
+{% set orca_state = 'off' if "planner" in test_name else 'on' %}
```
I understand how this does help avoid a little bit of pipeline yaml duplication but it would be the only job in the pipeline using such a pattern. Currently when you generate a pipeline the only internal sort of dimension you can split on is platform. Nothing exists in `gen_pipeline.py` for generating a pipeline based on `orca` or `planner`. Most of the templatizing that goes on in this template is to allow the different choices in `gen_pipeline.py`. So this bit of logic doesn't really impact the users choices when setting pipelines.

I would propose just dealing with the little bit of duplication in order to avoid creating a new layer of logic to be aware of in the pipeline templatization. If you team is ok with maintaining this then it's ok, but I caution against making that choice.

We are happy to chat about this more since it's difficult to articulate.

kaknikhil replied:
I do agree and I had decided on removing the `aa_Test_suites` variable. The only reason we kept it is because that's how it is set up in 5X_STABLE pipeline and we didn't want to touch the 5X pipeline. What do you think ?

kaknikhil replied:
On second thought, I agree that we should remove it.

kmacoskey commented on concourse/pipelines/templates/gpdb-tpl.yml:
```diff
@@ -2198,6 +2246,107 @@ jobs:
+      ORCA: "[[ orca_state ]]"
+      DBVER: gpdb6-assert
+
+{% if "centos7" in os_types %}
```
The templatizing exists to include/exclude `centos7` but not `centos6`. Is it intentional to make an assumption that `centos6` should always be included as a platform when generating a subset of the pipeline that includes the `AA` group? If it's not intentional, then the above job would need:

```
{% if "centos6" in os_types %}
- name: MADlib_Test_[[ test_name ]]_centos6
...
{% endif %}
```

kmacoskey replied:
There are a number of other places in the template where this decision would need to reflect as well if changed.

kaknikhil replied:
For the centos6 part, yes it is intentional that centos6 will always be present. I think we copied that pattern from other jobs in the pipeline.

kmacoskey commented on concourse/pipelines/gpdb_master-generated.yml:
```diff
@@ -2167,6 +2201,130 @@ jobs:
+  - aggregate:
+    - get: gpdb_src
+      passed:
+      - compile_gpdb_centos6
```
The pattern in every other test in the pipeline, that isn't core greenplum, is to have a passed condition on the `ICW` group of jobs per platform. I think there is merit in maintaining that pattern, because ICW passing for a given SHA is a strong requirement for any sort of release candidate. To the point that no external/internal component should be testing GPDB release candidates that have not passed ICW.

kaknikhil replied:
I couldn't find external tests that have a passed condition on the `ICW` group of jobs. Maybe I am missing something, can you point me to an example ?
